core/watcher/watchertest/watcher.go

```diff
@@ -44,3 +47,49 @@ func DirtyKill[T any](c *gc.C, w watcher.Watcher[T]) {
 		c.Logf("ignoring failed to close for watcher")
 	}
 }
+
+// WatcherC embeds a gocheck.C and adds methods to help
+// verify the behaviour of generic watchers.
+type WatcherC[T any] struct {
+	C       *gc.C
+	Watcher watcher.Watcher[T]
+}
+
+// NewWatcherC() returns a WatcherC[T].
+func NewWatcherC[T any](c *gc.C, w watcher.Watcher[T]) WatcherC[T] {
+	return WatcherC[T]{
+		C:       c,
+		Watcher: w,
+	}
+}
+
+// AssertNoChange verifies that no changes are received from the watcher during
+// a testing.ShortWait time.
+func (w *WatcherC[T]) AssertNoChange() {
+	select {
+	case actual, ok := <-w.Watcher.Changes():
+		w.C.Fatalf("watcher sent unexpected change: (%v, %v)", actual, ok)
+	case <-time.After(testing.ShortWait):
+	}
+}
+
+// AssertChange asserts the given changes was reported by the watcher,
+// but does not assume there are no following changes.
+func (w *WatcherC[T]) AssertChange(assertion func(c *gc.C, received []T) bool) {
+	var received []T
+	timeout := time.After(testing.LongWait)
+	for a := testing.LongAttempt.Start(); a.Next(); {
+		select {
+		case actual, ok := <-w.Watcher.Changes():
+			w.C.Logf("WatcherC Watcher.Changes() => %# v", actual)
+			w.C.Assert(ok, jc.IsTrue)
+			received = append(received, actual)
+			w.C.Logf("received %+v", received)
+			if assertion(w.C, received) {
+				return
+			}
+		case <-timeout:
+			w.C.Fatalf("watcher did not send change")
```

Comment:
**suggestion:** This error is misleading. In fact, the watcher could have send *some* changes, but not enough to validate the assertion.